File: services/api-ts/src/application/task-service.ts
import { createTask, DomainError, transition, type Status, type Task } from "../domain/task.ts";
import type { Clock, IdGenerator, TaskRepository } from "./ports.ts";

export interface TaskServiceDeps {
  readonly repository: TaskRepository;
  readonly clock: Clock;
  readonly ids: IdGenerator;
}

/** The use cases. Depends on the domain and on ports, never on an adapter: src/main.ts chooses those. */
export class TaskService {
  readonly #deps: TaskServiceDeps;

  constructor(deps: TaskServiceDeps) {
    this.#deps = deps;
  }

  async create(title: string): Promise<Task> {
    const task = createTask(this.#deps.ids.next(), title, this.#deps.clock.now());
    await this.#deps.repository.save(task);
    return task;
  }

  async get(id: string): Promise<Task> {
    const task = await this.#deps.repository.get(id);
    if (task === undefined) throw new DomainError("NOT_FOUND", "task not found");
    return task;
  }

  list(): Promise<readonly Task[]> {
    return this.#deps.repository.list();
  }

  /** Stores nothing when the domain refuses the move. */
  async transition(id: string, next: Status): Promise<Task> {
    const moved = transition(await this.get(id), next, this.#deps.clock.now());
    await this.#deps.repository.save(moved);
    return moved;
  }
}

File: services/api-py/src/api_py/application/task_service.py
"""The use cases. Depends on the domain and on its ports, never on an adapter: main.py chooses those."""

from __future__ import annotations

from dataclasses import dataclass

from api_py.application.ports import Clock, IdGenerator, TaskRepository
from api_py.domain.task import DomainError, Status, Task, create_task, transition


@dataclass(frozen=True, slots=True)
class TaskService:
    repository: TaskRepository
    clock: Clock
    ids: IdGenerator

    def create(self, title: str) -> Task:
        task = create_task(self.ids.next(), title, self.clock.now())
        self.repository.save(task)
        return task

    def get(self, task_id: str) -> Task:
        task = self.repository.get(task_id)
        if task is None:
            raise DomainError("NOT_FOUND", "task not found")
        return task

    def list(self) -> tuple[Task, ...]:
        return self.repository.list()

    def transition(self, task_id: str, nxt: Status) -> Task:
        """Stores nothing when the domain refuses the move."""
        moved = transition(self.get(task_id), nxt, self.clock.now())
        self.repository.save(moved)
        return moved
